internal/controller/numaflowcontrollerrollout_controller.go

```diff
@@ -0,0 +1,427 @@
+	ctx context.Context,
+	controllerRollout *apiv1.NumaflowControllerRollout,
+	namespace string,
+) error {
+	numaLogger := logger.FromContext(ctx)
+
+	if !controllerRollout.DeletionTimestamp.IsZero() {
+		numaLogger.Info("Deleting NumaflowControllerRollout")
+		if controllerutil.ContainsFinalizer(controllerRollout, finalizerName) {
+			controllerutil.RemoveFinalizer(controllerRollout, finalizerName)
+		}
+		return nil
+	}
+
+	// add Finalizer so we can ensure that we take appropriate action when CRD is deleted
+	if !controllerutil.ContainsFinalizer(controllerRollout, finalizerName) {
+		controllerutil.AddFinalizer(controllerRollout, finalizerName)
+	}
+
+	// apply controller
+	phase, err := r.sync(controllerRollout, namespace, numaLogger)
+	if err != nil {
+		return err
+	}
+
+	if phase != gitopsSyncCommon.OperationSucceeded {
+		return fmt.Errorf("sync operation is not successful")
+	}
+
+	controllerRollout.Status.MarkRunning()
+	return nil
+}
+
+// applyOwnershipToManifests Applies GitSync ownership to Kubernetes manifests, returning modified manifests or an error.
+func applyOwnershipToManifests(manifests []string, controllerRollout *apiv1.NumaflowControllerRollout) ([]string, error) {
+	manifestsWithOwnership := make([]string, 0, len(manifests))
+	for _, v := range manifests {
+		reference, err := kubernetes.ApplyOwnership(v, controllerRollout)
+		if err != nil {
+			return nil, err
+		}
+		manifestsWithOwnership = append(manifestsWithOwnership, string(reference))
+	}
+	return manifestsWithOwnership, nil
+}
+
+// TODO: share sync logic with `syncer`
+func (r *NumaflowControllerRolloutReconciler) sync(
+	rollout *apiv1.NumaflowControllerRollout,
+	namespace string,
+	numaLogger *logger.NumaLogger,
+) (gitopsSyncCommon.OperationPhase, error) {
+
+	// Get the target manifests
+	version := rollout.Spec.Controller.Version
+	manifest := r.definitions[version]
+
+	// Applying ownership reference
+	manifests, err := SplitYAMLToString([]byte(manifest))
+	if err != nil {
+		return gitopsSyncCommon.OperationError, fmt.Errorf("can not parse file data, err: %v", err)
+	}
+	manifestsWithOwnership, err := applyOwnershipToManifests(manifests, rollout)
+	if err != nil {
+		return gitopsSyncCommon.OperationError, fmt.Errorf("failed to apply ownership reference, %w", err)
+	}
+
+	targetObjs, err := toUnstructuredAndApplyAnnotation(manifestsWithOwnership, rollout.Name)
+	if err != nil {
+		return gitopsSyncCommon.OperationError, fmt.Errorf("failed to parse the manifest, %w", err)
+	}
+	numaLogger.Debugf("found %d target objects associated with Numaflow Controller version %s; versions defined:%+v", len(targetObjs), version, r.definitions)
+
+	var infoProvider kubeUtil.ResourceInfoProvider
+	clusterCache, err := r.stateCache.GetClusterCache()
+	infoProvider = clusterCache
+	if err != nil {
+		return gitopsSyncCommon.OperationError, err
+	}
+	liveObjByKey, err := r.stateCache.GetManagedLiveObjs(rollout.Name, targetObjs)
```

Comment:
note I have [this](https://github.com/numaproj/numaplane/issues/8) issue written - sort of relates to my comment above as well.

Reply:
we can do it seperately